libcst/metadata/name_provider.py

```diff
@@ -0,0 +1,166 @@
+        >>>         elif otherthing:
+        >>>             from d import e as c
+        >>>         else:
+        >>>             from f import g as c
+        >>>         c()
+        >>>     '''
+        >>>     ))
+        >>> )
+        >>> call = wrapper.module.body[1].body[0].value
+        >>> wrapper.resolve(QualifiedNameProvider)[call],
+        {
+            QualifiedName(name="a.b", source=QualifiedNameSource.IMPORT),
+            QualifiedName(name="d.e", source=QualifiedNameSource.IMPORT),
+            QualifiedName(name="f.g", source=QualifiedNameSource.IMPORT),
+        }
+
+    For qualified name of a variable in a function or a comprehension, please refer
+    :func:`~libcst.metadata.Scope.get_qualified_names_for` for more detail.
+    """
+
+    METADATA_DEPENDENCIES = (ScopeProvider,)
+
+    def visit_Module(self, node: cst.Module) -> Optional[bool]:
+        visitor = QualifiedNameVisitor(self)
+        node.visit(visitor)
+
+    @staticmethod
+    def has_name(
+        visitor: MetadataDependent, node: cst.CSTNode, name: Union[str, QualifiedName]
+    ) -> bool:
+        """Check if any of qualified name has the str name or :class:`~libcst.metadata.QualifiedName` name."""
+        qualified_names = visitor.get_metadata(QualifiedNameProvider, node, set())
+        if isinstance(name, str):
+            return any(qn.name == name for qn in qualified_names)
+        else:
+            return any(qn == name for qn in qualified_names)
+
+
+class QualifiedNameVisitor(cst.CSTVisitor):
+    def __init__(self, provider: "QualifiedNameProvider") -> None:
+        self.provider: QualifiedNameProvider = provider
+
+    def on_visit(self, node: cst.CSTNode) -> bool:
+        scope = self.provider.get_metadata(ScopeProvider, node, None)
+        if scope:
+            self.provider.set_metadata(node, scope.get_qualified_names_for(node))
+        else:
+            self.provider.set_metadata(node, set())
+        super().on_visit(node)
+        return True
+
+
+DOT_PY: Pattern[str] = re.compile(r"(__init__)?\.py$")
+
+
+def _module_name(root_path: Path, path: str) -> Optional[str]:
+    try:
+        return (
+            DOT_PY.sub("", str(Path(path).relative_to(root_path)))
+            .replace("/", ".")
+            .rstrip(".")
+        )
+    except ValueError:
+        # relative_to failed, we can't guess the module name
+        return None
+
+
+class FullyQualifiedNameProvider(BatchableMetadataProvider[Collection[QualifiedName]]):
+    """
+    Provide fully qualified names for CST nodes. Like :class:`QualifiedNameProvider`,
+    but the provided :class:`QualifiedName`s have absolute identifier names instead of
+    local to the current module.
+
+    This provider is initialized with the current module's fully qualified name, and
+    can be used with :class:`~libcst.metadata.FullRepoManager`. Compared to :class:`QualifiedNameProvider`, it also resolves relative imports.
+    """
```

Comment:
Can you add some explanation about the module name relative to the project root is saved as metadata of module node?
Also, add some code examples like what we did in QualifiedNameProvider docstring can make it much more clear to readers.